flintrock.py

```diff
@@ -526,42 +526,104 @@ def launch_ec2(
-            image_id=ami,
-            min_count=(num_slaves + 1),
-            max_count=(num_slaves + 1),
-            key_name=key_name,
-            instance_type=instance_type,
-            placement=availability_zone,
-            security_group_ids=[sg.id for sg in security_groups],
-            subnet_id=subnet_id,
-            placement_group=placement_group,
-            tenancy=tenancy,
-            ebs_optimized=ebs_optimized,
-            instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
-
-        time.sleep(10)  # AWS metadata eventual consistency tax.
-
-        # TODO: Move this to a reusable function and add a limit on
-        #       wait time.
-        while True:
-            for instance in reservation.instances:
-                if instance.state == 'running':
-                    continue
+        #spot instance request
+        if spot_price is not None:
+            print("Requesting {c} spot instances at a max price of {p}".format(c=machine_num, p=spot_price))
+            
+            instance_reqs = connection.request_spot_instances(
+                price=spot_price,
+                image_id=ami,
+                count=(machine_num), 
+                key_name=key_name, 
+                instance_type=instance_type, 
+                #launch_group="launch-group-%s" % cluster_name,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized,
+                
+                )
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+            # now we loop:
+            print("Waiting for spot instances to be granted...")
+            try:
+                while True:
+                    time.sleep(10)
+                    reqs = connection.get_all_spot_instance_requests() 
+
+                    id_to_req = {}
+                    for r in reqs:
+                        id_to_req[r.id] = r
+                    active_instance_ids = []
+                    for i in spot_instance_request_ids:
+                        if i in id_to_req and id_to_req[i].state == "active":
+                            active_instance_ids.append(id_to_req[i].instance_id)
+                    if len(active_instance_ids) == (machine_num):
+                        print("All {machine_num} granted".format(machine_num = machine_num))
+                        reservation_resultset = connection.get_all_reservations(active_instance_ids)
+
+                        all_instances = []
+                        for reservation in reservation_resultset:
+                            all_instances += reservation.instances
+                            
+                        break
+                    else:
+                        print("{grant} of {req} instances granted, waiting longer".format(
```

Comment:
I have found that it's nice to have this heartbeat -- sometimes it can take up to 20 min for a large spot instance to be granted, even if you're far above the current market price.